subprojects/packagefiles/luajit/src/meson.build

```diff
@@ -0,0 +1,274 @@
+    _target['output'],
+    input: _target['input'],
+    output: _target['output'],
+    command: [buildvm_exe, '-m', _target['mode'], '-o', '@OUTPUT@', '@INPUT@'],
+  )
+endforeach
+
+custom_target(
+  'vmdef.lua',
+  input: lib_sources,
+  output: 'vmdef.lua',
+  command: [buildvm_exe, '-m', 'vmdef', '-o', '@OUTPUT@', '@INPUT@'],
+  install: true,
+  install_dir: jitdir,
+  install_tag: 'runtime',
+)
+
+luajit_lib_flags = TARGET_XCFLAGS
+luajit_lib_flags += [
+  '-D_LARGEFILE_SOURCE',
+  '-U_FORTIFY_SOURCE',
+  ]
+
+luajit_lib_src = [buildvm_headers, lj_vm_out, luajit_h]
+if get_option('amalgam')
+  luajit_lib_src += 'ljamalg.c'
+else
+  luajit_lib_src += lib_sources
+  luajit_lib_src += [
+    'lib_aux.c',
+    'lib_init.c',
+    'lj_alloc.c',
+    'lj_api.c',
+    'lj_asm.c',
+    'lj_assert.c',
+    'lj_bc.c',
+    'lj_bcread.c',
+    'lj_bcwrite.c',
+    'lj_buf.c',
+    'lj_carith.c',
+    'lj_ccall.c',
+    'lj_ccallback.c',
+    'lj_cconv.c',
+    'lj_cdata.c',
+    'lj_char.c',
+    'lj_clib.c',
+    'lj_cparse.c',
+    'lj_crecord.c',
+    'lj_ctype.c',
+    'lj_debug.c',
+    'lj_dispatch.c',
+    'lj_err.c',
+    'lj_ffrecord.c',
+    'lj_func.c',
+    'lj_gc.c',
+    'lj_gdbjit.c',
+    'lj_ir.c',
+    'lj_lex.c',
+    'lj_lib.c',
+    'lj_load.c',
+    'lj_mcode.c',
+    'lj_meta.c',
+    'lj_obj.c',
+    'lj_opt_dce.c',
+    'lj_opt_fold.c',
+    'lj_opt_loop.c',
+    'lj_opt_mem.c',
+    'lj_opt_narrow.c',
+    'lj_opt_sink.c',
+    'lj_opt_split.c',
+    'lj_parse.c',
+    'lj_prng.c',
+    'lj_profile.c',
+    'lj_record.c',
+    'lj_serialize.c',
+    'lj_snap.c',
+    'lj_state.c',
+    'lj_str.c',
+    'lj_strfmt.c',
+    'lj_strfmt_num.c',
+    'lj_strscan.c',
+    'lj_tab.c',
+    'lj_trace.c',
+    'lj_udata.c',
+    'lj_vmevent.c',
+    'lj_vmmath.c',
+  ]
+endif
+
+luajit_lib = library(
+  'luajit-' + abiver,
+  luajit_lib_src,
+  c_args: luajit_lib_flags,
+  dependencies: [m_dep, dl_dep],
+  install: true,
+  override_options: ['b_sanitize=none'],
+  soversion: meson.project_version(),
+  # A version number like `2.1.1713773202` breaks clang when compiling for x86_64 on macOS:
+  # ```
+  # cc  -o […]/libluajit-5.1.2.1.1713773202.dylib […] -compatibility_version 2.1.1713773202 -current_version 2.1.1713773202 -lm -ldl
+  # ld: malformed 32-bit x.y.z version number: 2.1.1713773202
+  # clang: error: linker command failed with exit code 1 (use -v to see invocation)
+  # ````
+  # So use the same versions as in the official makefile.
+  darwin_versions: [ljver, ljver + '.255'],
+)
+
+luajit_dep = declare_dependency(
+  include_directories: include_directories('.'),
+  link_with: luajit_lib,
+  sources: luajit_h,
+  variables: {
+    'INSTALL_CMOD': install_cmod,
+    'INSTALL_LMOD': install_lmod,
+  },
+)
+meson.override_dependency('luajit', luajit_dep)
+
+if get_option('default_library') == 'static'
+  if host_machine.system() == 'windows'
+    error('building a static LuaJIT library is not supported on Windows')
```

Comment:
Also why is it unsupported? Actually this is the target I would be most interested in, static luajit.

Reply:
This is what the official Makefile has to say:
```make
  # Mixed mode is not supported on Windows. And static mode doesn't work well.
  # C modules cannot be loaded, because they bind to lua51.dll.
```

Reply:
Here are the official supported modes:
```make
##############################################################################
################################  BUILD MODE  ################################
##############################################################################
# The default build mode is mixed mode on POSIX. On Windows this is the same
# as dynamic mode.
#
# Mixed mode creates a static + dynamic library and a statically linked luajit.
BUILDMODE= mixed
#
# Static mode creates a static library and a statically linked luajit.
#BUILDMODE= static
#
# Dynamic mode creates a dynamic library and a dynamically linked luajit.
# Note: this executable will only run when the library is installed!
#BUILDMODE= dynamic
```

Reply:
I would change to `warning` if anything. I agree `mixed`(default_library=both) mode on Windows may/will not work. But dynamic(shared) or static works ok.

MSYS2 ships static luajit https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-luajit/PKGBUILD

I've also used staticky linked version. So I would keep it as an option with know limitations of Windows static linking.

Reply:
OK, so how about:
- if you're not building the luajit executable, use whichever library mode you like (including on Windows)
- otherwise on Windows:
  - `both`: unsupported, error
  - `shared`: OK
  - `static`: warning

?

Reply:
Ok for me.

Reply:
Done, please review.

Reply:
I've tested in my pipeline, builds and works correctly. Warning is little bit ominous, because except external modules there shouldn't be problems. Nevertheless it is fine.